### PR TITLE
add sentry error reporting

### PR DESCRIPTION
**Pull request by maks**
Add Sentry reporting for app across Flutter, native Android and iOS.

Includes refactor, renaming the App widget and moving it into its own file.

This has no impact on user visible UI or functionality.

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -74,4 +74,8 @@ flutter {
 
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+
+    implementation 'io.sentry:sentry-android:1.7.27'
```
Comment (workerbee22) [inline android/app/build.gradle:78]: So is this required to catch the Android platform level stuff? This seems to be an Android Sentry implementation?

File: android/app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -6,7 +6,7 @@
          additional functionality it is fine to subclass or reimplement
          FlutterApplication and put your custom class here. -->
     <application
-        android:name="io.flutter.app.FlutterApplication"
+        android:name=".WHApplication"
```
Comment (workerbee22) [inline android/app/src/main/AndroidManifest.xml:9]: Do we have a reason for changing this?

File: android/app/src/main/kotlin/org/whegba/wh_covid19/WHApplication.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.whegba.wh_covid19
```
Comment (workerbee22) [inline android/app/src/main/kotlin/org/whegba/wh_covid19/WHApplication.kt:1]: Kotlin code? Again I assume this is to implement the Android platform side of it?

File: ci_cd.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+### Sentry config
```
Comment (workerbee22) [inline ci_cd.md:1]: Assume this goes along with codemagic changes? Which script in codemagic? Make sure you don't clobber the Post script we have atm for the Firebase App Dist things.

File: lib/error_reporting.dart [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import 'package:device_info/device_info.dart';
+import 'package:flutter/foundation.dart';
+import 'package:package_info/package_info.dart';
+import 'package:sentry/sentry.dart';
+import 'env.dart';
+
+Future<SentryClient> get _sentryClient async => SentryClient(
+      dsn: dsn,
+      environmentAttributes: await _environmentEvent,
+    );
```
Comment (workerbee22) [inline lib/error_reporting.dart:13]: dartfmt formatted?

File: lib/error_reporting.dart [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      version: androidInfo.version.release,
+    );
+    device = Device(
+      model: androidInfo.model,
+      manufacturer: androidInfo.manufacturer,
+      modelId: androidInfo.product,
+    );
+  } else if (Platform.isIOS) {
+    final iosInfo = await deviceInfoPlugin.iosInfo;
+    os = OperatingSystem(
```
Comment (workerbee22) [inline lib/error_reporting.dart:33]: Good these are the best values for iOS 👍

File: lib/error_reporting.dart [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      name: iosInfo.systemName,
+      version: iosInfo.systemVersion,
+    );
+    device = Device(
```
Comment (workerbee22) [inline lib/error_reporting.dart:37]: ditto 👍

File: lib/error_reporting.dart [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        name: packageInfo.appName,
+        version: packageInfo.version,
+        build: packageInfo.buildNumber,
+      ),
+    ),
+  );
+  return environment;
+}
+
+void flutterOnErrorHandler(FlutterErrorDetails details) {
```
Comment (workerbee22) [inline lib/error_reporting.dart:59]: Cool uncaught errors 👍

File: lib/main.dart [hunk 9]
```diff
@@ -1,137 +1,37 @@
...
-import 'view/ppe/ppe_view.dart';
-import 'view/sbs_guide_view.dart';
-import 'view/staff_welfare/your_welfare_view.dart';
+import 'error_reporting.dart';
+import 'wh_app.dart';
 
 void main() {
   // Add this, and it should be the first line in main method to ensure no crashes before runApp()
-  WidgetsFlutterBinding.ensureInitialized();
+  //WidgetsFlutterBinding.ensureInitialized();
```
Comment (workerbee22) [inline lib/main.dart]: Do we need to comment this out? Does it cause issues? I've found it helps prevent ugly issues at app startup.

File: lib/main.dart [hunk 10]
```diff
@@ -1,137 +1,37 @@
...
 
-  runApp(const MyApp());
-}
+  // This captures errors reported by the Flutter framework.
+  FlutterError.onError = flutterOnErrorHandler;
+
+  // Lets not show end users the red screen of death
```
Comment (workerbee22) [inline lib/main.dart:16]: Yep ...

**Review by workerbee22 — APPROVED** [review]
LGTM if a bit more complicated than I'm used to for Sentry.

**Comment by workerbee22** [comment]
@maks There is no more iOS setup I can do quickly to trap iOS platform errors, like you have here for Android. So no more commits from me on this for iOS right now.

File: android/app/build.gradle (hunk 1 above)
Comment (maks) [reply android/app/build.gradle:78]: Yes we want to be able to catch errors that happen in native Android code, either in Flutter framework or plugins, which won't necessarily be reported to Dart.

File: android/app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (maks) [reply android/app/src/main/AndroidManifest.xml:9]: Yes, we need a custom Android App singleton class in order to do native android Sentry initialisation in its `onCreate()`

File: android/app/src/main/kotlin/org/whegba/wh_covid19/WHApplication.kt (hunk 3 above)
Comment (maks) [reply android/app/src/main/kotlin/org/whegba/wh_covid19/WHApplication.kt:1]: yep see my comment above about native Android sentry init

File: ci_cd.md (hunk 4 above)
Comment (maks) [reply ci_cd.md:1]: Ah yes I should include here that it needs to be a `pre-build` step Codemagic workflow script

File: lib/main.dart (hunk 9 above)
Comment (maks) [reply lib/main.dart]: Oops, I thats a mistake, I meant to put that back.

File: android/app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (workerbee22) [reply android/app/src/main/AndroidManifest.xml:9]: Yeah cool 👍

**Pull request merged**